// io/parsers.js
// SPDX-FileCopyrightText: 2025 Karsten Teichmann
//
// SPDX-License-Identifier: MIT

/**
 * @file
 * @description
 * This file contains a set of very simple parsers providing just enough of the specs/grammars 
 * to handle the different property file languages and styles used in batocera.linux:
 * - es_systems.yml
 * - es_features.yml
 * - configgen-*.yml
 * - *.conf files
 * - *.cfg files
 * - *.json
 * - *.xml
 * 
 * For XML and YAML formatted files, an external tool will be used (xq and yq) to transform to json first. 
 * This makes those types effectively twice as expensive to process.  
 * Thus, where speed matters, the user should not use yml. Moreover, frequently parsed files should be handled with 
 * dedicated in-process parsing functions.  
 * Example: `es_settings.cfg` is an xml file, but very simply structured.
 */
const log = require('logger').get()
const data = require('utils/data');

const path = require('node:path');
const { readFileSync, existsSync } = require('node:fs');
const { execSync } = require('node:child_process')

const SOURCE_FILE = Symbol.for('#SOURCE');
let CURRENT_FILE;

const PARSE_FUNCTIONS = Object.freeze({
  'yml': yamlToDict,
  'yaml': yamlToDict,
  'conf': confToDict,
  'json': jsonToDict,
  'cfg': esSettingsToDict,
  'xml': xmlNew,
  'amgp': xmlNew
});

const CONTENT_PROVIDER = Object.freeze({
  DIRECT: (filename, split = true) => {
    let result = existsSync(filename)
      ? readFileSync(filename, { encoding: 'utf8' })
      : filename;
    if (split) { result = result.split('\n') }
    return result;
  },
  YQ: (filename, split = true) => {
    return existsSync(filename)
      ? execSync(`yq . "${filename}"`, { encoding: 'utf8' })
      : execSync('yq', { encoding: 'utf8', input: filename });
  },
  XQ: (filename, split = true) => {
    return existsSync(filename)
      ? execSync(`xq . "${filename}"`, { encoding: 'utf8' })
      : execSync('xq', { encoding: 'utf8', input: filename });
  }
});

/**
 * This function is the main entrance to this file.  
 * It tries to auto-detect the parser function to use depending on the file type.  
 * The detection is rudimentary, it goes by extension. Concrete handler functions are looked up
 * in a constant object named `PARSE_FUNCTIONS`.
 * 
 * All handler functions follow the same contract:
 * - They will generally return nested object structures.
 * - Each actual property entry will we wrapped in an instance of `PropValue`.  
 * 
 * @exported
 * @see class PropValue
 */
function parseDict(confFile, typeOverride = null) {
  let fileType;
  let isPropertyString = false;
  if (existsSync(confFile)) {
    fileType = path.extname(confFile).substring(1);
  } else if (typeof confFile == "string") {
    fileType = 'conf';
    isPropertyString = true;
  }

  let effectiveType = typeOverride || fileType;
  let parser = PARSE_FUNCTIONS[effectiveType];

  if (isPropertyString) {
    //a string, but not a file path -> assume '.'-imploded property keys
    log.debug(`Parsing property string as [${effectiveType}] format`)
    return parser(confFile);
  }

  try {
    if (typeof parser == "undefined") { throw new Error('unsupported file type/extension: ' + confFile) }
    let resultDict = parser(confFile);
    resultDict[SOURCE_FILE] = confFile;
    return resultDict;
  } catch (e) {
    log.error("Could not parse file:", confFile, e);
    return {}
  }
}

/**
 * Handles `*.conf` files like `batocera.conf`.  
 * Capable of unpacking batocera's special syntaxes into regular tree structures:
 * - system["<game-name>"].something=...
 * - system.folder["<folder-path>"].something=...
 * 
 * @exported
 * @see analyseProperty
 */
function confToDict(confFile) {
  let assign = data.deepAssign;
  return readTextPropertyFile(confFile, lines => {
    let result = {};
    for (let line of lines) {
      let propLine = line.trim();
      if (propLine.startsWith('#') || !propLine.includes("=")) { continue; }
      let details = analyseProperty(propLine);
      assign(result, details.effectiveKey, details.value);
    }
    return result;
  })
}

const XML = {
  ENCODED_CHARS_REGEX: /&lt;|&gt;|&amp;|&apos;|&quot;/g,
  removeComments(lines) {
    if (Array.isArray(lines)) { lines = lines.join('\n') }

    while (lines.includes('<!--')) {
      let startIndex = lines.indexOf('<!--');
      let endIndex = lines.indexOf('-->', startIndex);
      lines = lines.substring(0, startIndex) + lines.substring(endIndex + 3);
    }
    return lines.split('\n');
  },

  condenseWhitespace(value) {
    return value.replace(/\s*$\s*/gm, ' ').trim();
  },

  decodeValue(value) {
    return value.replace(XML.ENCODED_CHARS_REGEX, match => {
      switch (match) {
        case '&lt;': return '<'
        case '&gt;': return '>'
        case '&amp;': return '&'
        case '&apos;': return "'"
        case '&quot;': return '"'
      }
    });
  },

  decodeAttribute(value) {
    return XML.decodeValue(XML.condenseWhitespace(value));
  },
}

const CFG_PROP_LINE = /<(\w+) name="(.*)" value="(.*)"\s*\/>/
const SETTINGS_TYPES = {
  string: String,
  float: Number,
  int: Number,
  bool: Boolean
}
function esSettingsToDict(cfgFile) {
  return readTextPropertyFile(cfgFile, (lines) => {
    lines = XML.removeComments(lines);
    let result = {};
    lines.map(_ => CFG_PROP_LINE.exec(_)).filter(_ => _ != null).forEach(line => {
      let key = XML.decodeValue(line[2]);
      let convertedProperty = `${key}=${XML.decodeValue(line[3])}`;
      let details = analyseProperty(convertedProperty);
      details.effectiveKey.set(result, SETTINGS_TYPES[line[1]](details.value.valueOf()));
    });
    return result;
  });
}

function jsonToDict(jsonFile, contentProvider = CONTENT_PROVIDER.DIRECT) {
  let isNativeJson = contentProvider == CONTENT_PROVIDER.DIRECT;

  function noComment(line) { return !/\s*\/\/.*/.test(line) }
  function propertyNodeCreator(key, value) {
    if (typeof value == "object" && value != null) { return value }
    else { return handleValue(String(value)) }
  }
  return readTextPropertyFile(jsonFile, (lines) => {
    let data = isNativeJson
      ? lines.filter(noComment).join('\n')
      : lines;
    try {
      return JSON.parse(data, propertyNodeCreator);
    } catch (e) {
      log.error(`Parsing [${jsonFile}] failed. Content was:`, data);
      log.error(e);
      return {};
    }
  }, contentProvider, isNativeJson);
}

function xmlToDict(xmlFile) {
  let result = jsonToDict(xmlFile, CONTENT_PROVIDER.XQ);
  return typeof result['NO-ROOT'] != "undefined"
    ? result['NO-ROOT']
    : result;
}

function xmlNew(xmlFile) {
  let result = readTextPropertyFile(xmlFile, sourceText => {
    let diag = false;

    let context = new ParseContext();
    context.addAttributeNode = function(m) {
      this.enterNode(`@${m[1]}`, m[1] + '=');
      this.addContent(handleValue(XML.decodeAttribute(m[2].slice(1, -1))), m[2])
      this.leaveNode();
    }.bind(context);

    let R = new RegexDocParser({
      // none-node whitespace
      BLK: [/\s+/ms, m => context.current.addRaw(m[0])],
      // start of node
      SON: [/<(?!\/)([\p{L}\._\d-]+?)([\s>]|(?=\/))/su, m => context.enterNode(m[1], m[0])],
      //attribute
      NATTR: [/(\w+)=("[^"]*"|'[^']*')/s, context.addAttributeNode],
      // end of node opening tag
      EOH: [/\s*\>/ms, m => context.nodeHeaderDone(m[0])],
      // node value/content
      VAL: [/(\S[^\<\>]*?)\s*(?=\<)/ms, m => context.addContent(handleValue(XML.decodeValue(m[1])), m[0])],
      // end of the node (closing tag)
      EON: [/<?\/[\p{L}\._\d-]*?>/msu, m => context.leaveNode(m[0])],
    }, diag);

    sourceText = sourceText.replace(/\<\?.*?\?\>/gs, '');
    sourceText = sourceText.replace(/\<!--.*?--\>/gs, '');

    context.start();
    R.exec(sourceText);

    return context.current.valueOf();
  }, CONTENT_PROVIDER.DIRECT, false);
  return typeof result['NO-ROOT'] != "undefined"
    ? result['NO-ROOT']
    : result;
}

function yamlToDict(yamlFile) {
  return jsonToDict(yamlFile, CONTENT_PROVIDER.YQ);
}

/**
 * Utilities to handle the complex parser state when parsing yaml.
 * Methods of this object will be called with `this` set to the instance of the current `ParseContext`.
 */
class ParseContext {
  #STACK = []

  get root() { return this.#STACK[0]; }
  get current() { return this.#STACK.at(-1); }
  get previous() { return this.#STACK.at(-2); }

  /** 
   * `this` is normally lost when methods are used as callbacks.
   * This method overrides that by assigning a bound variant of the class method as instance property.
   * Passing the instance property as callback means passing a pre-bound function.
   */
  bindFunctions() {
    Object.values(Object.getOwnPropertyDescriptors(Object.getPrototypeOf(this))).forEach(v => {
      v = v.value;
      if (typeof v == "function") { this[v.name] = v.bind(this) }
    });
  }

  start(parser) {
    this.parser = parser;
    this.#STACK = [new UniversalNode('#root')];
  }

  enterNode(name, raw = name) {
    let node = new UniversalNode(name, raw);
    this.current.subNode(name, node);
    this.#STACK.push(node);
    console.error('\t--> ENTER', this.current.tagName())
  }

  addChildNode(tag) {
    let node = new UniversalNode(tag);
    this.current.content(node)
    this.#STACK.push(node);
    console.error('\t--> PUSH CHILD:', tag)
  }

  nodeHeaderDone(raw) { this.current.headerDone(raw); }

  // leave node
  leaveNode(raw) {
    console.error('\t--> LEAVE', this.current.tagName(), this.current)
    this.current.addRaw(raw, true);
    this.#STACK.pop();
  }

  addContent(text, raw) { this.current.content(text, raw); }
}

class YAML extends ParseContext {
  // PARSE_MODES is not actively used (referenced in the code)
  static PARSE_MODES = ['BLOCK', 'FLOW']
  // something like a 'sub-category' of parse modes
  static PARSE_PHASES = ['ENTRY', 'TOKEN', 'TOKEN:ML']
  static SCALAR_STYLE_OPS = { '|': "LITERAL", '>': "FOLD" }

  static CNEST = Symbol.for('CNEST');
  static NESTING = Symbol.for('NESTING');
  static MODE = Symbol.for('MODE');
  static SCALAR_STYLE = Symbol.for("SCALAR");
  static PHASE = Symbol.for('PHASE');
  static LAST_TOKEN = Symbol.for('LASTTOKEN');

  cleanNode() { Object.getOwnPropertySymbols(this.current).forEach(k => delete this.current[k]); }
  pushNextNode(tagName, isSubNode) {
    if (typeof isSubNode == "undefined") { isSubNode = tagName ? String(tagName).length > 0 : false; }
    let factory = isSubNode ? this.enterNode : this.addChildNode;
    factory.call(this, tagName);
    this.inheritStateProperties();
    this.current[YAML.PHASE] = "ENTRY";
    delete this.previous[YAML.PHASE];
  }
  inheritStateProperties() {
    this.current[YAML.MODE] = this.previous[YAML.MODE] || 'BLOCK';
    this.current[YAML.NESTING] = this.previous[YAML.CNEST];
    this.current[YAML.SCALAR_STYLE] = this.previous[YAML.SCALAR_STYLE] || "FOLD";
    console.error(
      `\t--> INHERIT ${this.current.tagName()}:`, this.current[YAML.NESTING], '->', this.current[YAML.MODE],
      `from:`, this.previous.tagName(), this.previous[YAML.NESTING], this.previous[YAML.CNEST]
    );
  }

  hasToken() { return Array.isArray(this.state(YAML.LAST_TOKEN)); }
  handleToken(match) {
    // when coming from parser directly, the first group will be token content
    // otherwise it could be called from another method with a pre-processed string
    match = Array.isArray(match) ? match[1] : match;
    (this.current[YAML.LAST_TOKEN] ||= []).push(match);
    let content = this.state(YAML.LAST_TOKEN).join("");
    let hasContent = content.trim().length > 0;
    switch (this.state(YAML.PHASE)) {
      case "ENTRY":
        if (hasContent) {
          this.state(YAML.PHASE, "TOKEN");
          this.stateOnce(YAML.SCALAR_STYLE, "FOLD");
        }
        break;
      case "TOKEN":
        this.state(YAML.PHASE, "TOKEN:ML");
        break;
    }
  }
  consumeToken() {
    if (!this.hasToken()) { return; }

    let result = this.condenseToken();
    delete this.current[YAML.LAST_TOKEN];
    delete this.current[YAML.PHASE];
    return result;
  }

  /** For keys and values: handles trimming, folding, chomping etc. */
  condenseToken() {
    console.error("\t--> CONDENSE:", this.current[YAML.LAST_TOKEN])
    let result = this.current[YAML.LAST_TOKEN].join('');
    if (result.trim().length == 0) { return; }

    let repeater = '+';
    // determine if (and how many) indentation/nesting spaces must be removed
    // FIXME: cant use PHASE for this - conflicts with TOKEN:ML.
    // TODO: introduce some 'config' subobject for SCALAR_STYLE instead of just using string constants
    switch (this.state(YAML.PHASE)) {
      case 'SCALAR:FL': // flow style
        break;

      case 'SCALAR:BL': // block style
        let nesting = this.state(YAML.NESTING) || 0;
        nesting += 0
        break;
    }
    result = result.replace(r`/(?<=\n)[ \t]${repeater}(?=\S)/g`, '');

    //TODO: adjust leading spaces based on nesting
    if (this.state(YAML.SCALAR_STYLE) == "FOLD") {
      result = result
        .replace(/(?<=\n)[ \t]+(?=\S)/g, '')
        .replace(/(?<=\S)[ \t]+(?=\n)/g, '')
        .replace(/(?<=\S)\n(?=\S)/g, ' ')
        .replace(/(?<=\S)\n(\n+)/g, '$1')
        .trim();
    } else {
      let nesting = this.state(YAML.CNEST) || 0;
      result = result
        //remove number of characters to matching the current nesting.
        .replace(r`/(?<=\n|^)[ \t]{${nesting}}/g`, '');
    }
    return unquote(result.trim());
  }
  tokenToContent() {
    let token = this.consumeToken();
    if (typeof token == "undefined") { return }

    console.error('\t--> ASSIGN TOKEN', token);
    this.addContent(handleValue(token));
  }

  enterFlow(flowNodeMarker) {
    this.pushNextNode('#FLOW:ROOT', false);
    this.current[YAML.MODE] = 'FLOW';
    if (flowNodeMarker[0].trim() == "[") { this.current.setContentMode("ARRAY"); }
    if (this.current.getContentMode() == "ARRAY") {
      this.pushNextNode('#ARR-ENTRY', false);
    }
  }

  closeFlow(match) {
    let term = match[0].replace(/\s+/gm, '');
    if (term == ',') {
      this.current.getContentMode() == "ARRAY" ? this.tokenToContent() : this.finishCurrentNode();
      return;
    }

    this.finishCurrentNode();
    if (!this.previous) { return; }
    console.error('\t    MODES (p,c)', this.previous[YAML.MODE], this.current[YAML.MODE])
    if (this.current.tagName() == "#FLOW:ROOT") { this.finishCurrentNode(); }
  }

  handleOperator(op, reg, source) {
    op = op[0];
    switch (op.charAt(0)) {
      case ':':
        let token = this.consumeToken();
        this.pushNextNode(token);
        break;
      case '|':
      case '>':
        this.state(YAML.SCALAR_STYLE, YAML.SCALAR_STYLE_OPS[op.charAt(0)]);
      //this.state(YAML.PHASE, "SCALAR:BL");
    }
  }

  markArray(depth, addChild = true) {
    depth = Number(depth) || 2;
    console.error("\t--> MARK ARRAY:", this.current.tagName(), 'N:', this.state(YAML.NESTING), 'CN:', this.current[YAML.CNEST]);
    if (this.current.getContentMode() != "ARRAY") {
      this.pushNextNode(this.current.tagName() + '[]', false);
      this.state(YAML.NESTING, this.previous[YAML.NESTING]);
      this.state(YAML.CNEST, this.previous[YAML.CNEST]);
      this.current.setContentMode("ARRAY");
    }

    if (addChild) {
      this.pushNextNode(this.current.tagName() + '#' + this.current.getContentLength(), false);
      this.state(YAML.CNEST, this.state(YAML.NESTING) + depth);
    }
  }

  finishCurrentNode() {
    this.tokenToContent();
    this.cleanNode();
    this.leaveNode();
  }

  state(state, value, node) {
    node ||= this.current;
    if (value == undefined) { return node[state]; }
    node[state] = value;
    console.error(`\t--> ${Symbol.keyFor(state)} ${value} >>> ${node.tagName()}`)
  }

  /** Only SET value if state is not set already */
  stateOnce(state, value, node) {
    node ||= this.current;
    if (typeof value == "undefined") { throw "stateOnce() must receive a value argument" }
    if (typeof node[state] == "undefined") { this.state(state, value, node); }
  }

  closeNodes(depthMatch) {
    let regD = depthMatch[1].length;
    let arrD = regD;
    if (depthMatch[0].trim().endsWith('-')) {
      arrD += 2;
      this.parser.lastIndex -= 2;
    }
    while (true) {
      if (arrD > regD && arrD > this.state(YAML.NESTING) && "ENTRY" == this.state(YAML.PHASE)) {
        this.state(YAML.CNEST, regD);
        this.markArray(arrD - regD, false);
      }
      let diff = this.state(YAML.NESTING) - (this.current.getContentMode() == "ARRAY" ? arrD : regD);
      console.error("\t--> RCL:" + this.current.tagName(), 'N:', this.current[YAML.NESTING], 'D:', diff, regD, arrD)

      if (diff >= 0) { this.finishCurrentNode(depthMatch); }
      else { break; }
    }
    this.stateOnce(YAML.CNEST, arrD);
    //save whitespace for unquoted multiline Strings
    this.handleToken(depthMatch[0].substring(0, regD + 1));
  }
}

function yamlToDict_IMPL(yamlFile) {
  return readTextPropertyFile(yamlFile, sourceText => {

    let yaml = new YAML();
    yaml.bindFunctions();

    function isMode(m, isIncluded = true) { return () => m.includes(yaml.current[YAML.MODE]) == isIncluded; }
    function isPhase(p, isIncluded = true) { return () => p.includes(yaml.current[YAML.PHASE]) == isIncluded; }

    let R = new RegexDocParser({
      // NOOP: Comments - (at least) one whitespace followed by whatever up to the next \n.
      // only allowed while not in the middle of a multiline token, treat it as regular content otherwise
      LC: [/(?:^|\s)+#.*?(?=$)/m, false, isPhase(["TOKEN:ML"], false)],
      // Empty lines
      EL: [/\n[ \t]*(?=\n|$)/m, yaml.handleToken],
      'BL:ARR': [/(-)(?=\s)/, yaml.markArray, isMode('BLOCK')],
      // From " on anything up until the first unescaped " (unescaped: no uneven number of slashes before it)
      // can run over multiple lines.
      DQ: [/("(?:\\\\|\\"|\\[^"]|[^\\"])*")/m, yaml.handleToken],
      // single quoted tokens
      SQ: [/('(?:''|[^'])*')/m, yaml.handleToken],
      // block scalar indicators
      'BL:SLH': [/([|>]|[|>][+-]\d|[|>]\d[+-])(?=[ \t\n])/, yaml.handleOperator, isMode('BLOCK')],
      // indicators which are not allowed in unquoted tokens
      OP: [/([:?](?=[ \t\n]))/, yaml.handleOperator],
      // indicators/operators belonging to flow mode: [], {}, ','
      'FL:OP': [/\{|\[/, yaml.enterFlow],
      'FL:X': [/\s*(?:,[ \t]*\}|,[ \t]*\]|[,\}\]])/, yaml.closeFlow, isMode('FLOW')],
      // unquoted in FLOW mode has additional terminators `,`, `}` and `]`
      'NQ:FL': [/(\S.*?)(?=[:?-]\s|[\]\},]|\s#|\n)/, yaml.handleToken, isMode('FLOW')],
      // any unquoted one-liner, up to the first indicator or linebreak
      // note: might have issues at end of file - requires an additional space or \n
      NQ: [/(\S.*?)(?=[:?-]\s|\s#|\n|$)/, yaml.handleToken],
      // the whitespace at the beginning of a line which controls when to leave a nested level
      DPT: [/(?:^|\n)([ \t]*)(-\s)?(?=[^\-\n]|\S)/m, yaml.closeNodes, isMode('BLOCK')],
      // In FLOW, only take the leading whitespaces as parts of the token, they are not used for closing.
      'DPT:FL': [/(\n\s*)/, yaml.handleToken, isMode('FLOW')],
      // NOOP: trailing whitespace from payload up to the next \n
      TRAIL: [/[ \t]+(?=\S|\n|$)/m, () => { }]
    }, true);

    yaml.start(R);
    yaml.current[YAML.NESTING] = -1;
    yaml.current[YAML.MODE] = 'BLOCK';
    try {
      R.exec(sourceText);
      yaml.closeNodes(["", ""]);
      yaml.cleanNode();

    } finally {
      console.error(JSON.stringify(yaml.root.valueOf(), null, 2))
    }
    return yaml.current.valueOf();
  }, CONTENT_PROVIDER.DIRECT, false);
}

const FOLDER_SPEC = /^\w+(\.folder\[(".*?")\/?\]).*=.*/;
const GAME_SPEC = /^\w+\[(".*?")\].*/;
function analyseProperty(propLine) {
  let fspath = []
  if (fspath = FOLDER_SPEC.exec(propLine)) {
    propLine = propLine.replace(fspath[1], '');
    fspath = ['folder', unquote(fspath[2])];
  } else if (fspath = GAME_SPEC.exec(propLine)) {
    fspath = fspath[1];
    propLine = propLine.replace(`[${fspath}]`, '');
    fspath = ['game', unquote(fspath)];
  } else {
    fspath = [];
  }

  let equalPos = propLine.indexOf('=');
  let realKey = propLine.substring(0, equalPos).trim().split('.');
  let effKey = [...realKey]

  return {
    effectiveKey: new data.HierarchicKey(effKey.shift(), ...fspath, ...effKey),
    overridesKey: new data.HierarchicKey(...realKey),
    value: handleValue(propLine.substring(equalPos + 1).trim())
  }
}

/**
 * @section Generic parser infrastructure
 * @description
 * The following functions and classes provide basic plumbing and utility for all type-specific parsing implementations.
 */

/**
 * A configurable generic parser based on regular expressions.  
 * The expressions will be set up to be sticky and share the same value for `lastIndex`.
 * Any successful match of one of the sub-expressions will advance the shared state,
 * but only when the match starts at the current `lastIndex`. It is not allowed to implicitely
 * skip and ignore parts of the input.
 * This allows to parse a long text without having to manipulate the input all the time 
 * which greatly improves speed. There is also no need to split line by line.  
 * 
 * Usage:
 * ```js
 * let ctx = new ParseContext();
 * let p = new RegexDocParser({
 *|  name : [ /pattern/flags, matchFunction(m){ ctx.doSomething(m) ... } ]
 * }, debugMatches)
 * 
 * p.exec(fullSource);
 * 
 * ```
 */
class RegexDocParser extends RegExp {
  #id;
  #diagMessages;
  #state;
  #rules = false;
  #ruleIds = false;
  static #patchFlags(reg, flags) {
    flags ||= reg.flags;
    if (!flags.includes('y')) { flags += 'y'; }
    if (!flags.includes('d')) { flags += 'd'; }
    return flags;
  }
  static #escapeSpaces(str) { return str.replace(/\n/g, '\\n').replace(/\t/g, '\\t'); }
  static #execRule(reg, action, useAllowed, sourceText, diag) {
    if (!useAllowed()) { return false; }

    let m = reg.exec(sourceText, diag);
    if (m == null) { return false; }

    let result = action
      ? action(m, reg, sourceText) || true
      : true;
    this.previousMatch = m;
    return result;
  }
  static #traverseSource(shared, sourceText) {
    let ruleIdx = 0;
    let lastRule = shared.#rules.length - 1;
    while (shared.lastIndex < sourceText.length) {
      let ruleResult = shared.#rules[ruleIdx](sourceText, shared.#diagMessages);
      if (ruleResult) { ruleIdx = shared.#ruleIds[ruleResult] || 0; }
      else { ruleIdx++; }

      if (ruleIdx > lastRule) {
        let snippet = sourceText.slice(shared.lastIndex, shared.lastIndex + 20);
        throw 'No valid expression for [' + JSON.stringify(snippet) + '...]';
      }
    }
  }
  constructor(state, id, a, b) {
    if (arguments.length <= 2) {
      super(/.*/);
      this.#rules = [];
      this.#ruleIds = {};
      Object.keys(state).forEach(k => {
        if (this[k]) { throw `Duplicate pattern name: ${k} (or reserved word)`; }
        this[k] = new RegexDocParser(this, k, state[k][0]);
        this.#ruleIds[k] = this.#rules.length;
        this.#rules.push(RegexDocParser.#execRule.bind(this, this[k], state[k][1], state[k][2] || (() => true)));
      });
      this.exec = function exec(str) { return RegexDocParser.#traverseSource(this, str); }.bind(this);
      this.#diagMessages = id || false;
      this.lastIndex = 0;
      return this;
    }

    super(a, RegexDocParser.#patchFlags(a, b));
    this.#state = state;
    state.reset ||= this.reset.bind(this);
    this.#id = id;
  }

  reset() { this.#state.lastIndex = 0; }
  exec(sourceText, diag) {
    let currentLast = this.#state.lastIndex;
    this.lastIndex = currentLast;
    let match = super.exec(sourceText);
    diag = !['TRAIL'].includes(this.#id);
    if (diag) {
      let snippet = RegexDocParser.#escapeSpaces(sourceText
        .substring(currentLast, currentLast + 30)
      ).substring(0, 32).padEnd(35, '.');
      match && console.error(
        `[${this.#id.padStart(6)}] `,
        `${snippet}`,
        "=>",
        '[' + RegexDocParser.#escapeSpaces((match || []).join('|')) + ']'
      );
    }
    if (match == null) { return null; }
    if (match.indices[0][0] != currentLast) {
      diag && console.error(this.#id, "did not start at lastIndex=" + currentLast, ", but at", this.lastIndex);
      return null;
    }

    this.#state.lastIndex = this.lastIndex;
    return match;
  }
  toString() { return super.toString(); }
}

/**
 * To be used in conjunction with [class RegexDocParser](#class_regexdocparser).
 * 
 * This class represents one node of any type in the parsed tree. Its basic capabilities are inspired by the structure produced by `xmlToDict`.
 * Depending on how it is populated during parsing, its `valueOf()` method will return different objects:
 * Either simple scalar values, dictionaries or arrays.  
 * This implementation will also recurse into children and subNodes to ensure proper mapping of the entire graph.
 *
 * `UniversalNode` internally manages 3 containers where content can be added:
 * - `#subNodes = {}`: Populated by `attr(name, value)` and `subNode(tag, node)`.
 *   Used for key-value tree structures. Like in `xmlToDict`, the same key can be used multiple times, `#subNodes` maintains each key's containt as array.
 * - `#content = []`: For lists or otherwise mixed content which is not (or shall not) be added as a property of this instance.
 * - `#rawValues = []`: This is used in all content adding methods in addition to either `#subNodes` or `#content`
 *   to maintain an unmodified view of the literal input source. Required for some edge cases of mixed content (specific to XML).
 *
 * Result:  
 * Usually, only one of either `#subNodes` or `#content` will be populated and `UniversalNode` is able to decide on its own what to return.  
 * Refer to [valueOf()](#class_universalnode_valueof) for details.
 */
class UniversalNode {
  #tagName;
  #rawValues = []
  #rawContentIndices = []
  #subNodes = {}
  #content = []
  #contentMode = 'DYNAMIC';

  static realValue(obj) { return (obj instanceof UniversalNode) ? obj.valueOf() : obj; }

  constructor(tag, raw) {
    this.#tagName = tag;
    this.#rawValues.push(raw);
  }

  /** 
   * ContentMode allows to control how the children and subnodes of this will be added when `getContent()` is used.  
   * Must be one of DYNAMIC, ARRAY or TEXT.
   */
  setContentMode(cm) { this.#contentMode = cm; }
  getContentMode() { return this.#contentMode; }

  // some small getter utils
  tagName() { return this.#tagName; }
  hasContent() { return this.#content.length > 0; }
  getContentLength() { return this.#content.length; }
  isMixed() { return this.hasContent() && Object.keys(this.#subNodes).length > 0 }
  toString() { return this.#rawValues.join(''); }

  // methods to manipulate the content of the instance
  /** 
   * Add an 'attribute' to the node. Attributes are similar to subNodes, with 2 differences:
   * - The name given will automatically be prefixed with `@`.
   * - Attributes do not allow several items. Every assignment to the same attribute will overwrite the current value.
   */
  attr(name, value, raw) {
    this.#rawValues.push(raw);
    this.#subNodes[`@${name}`] = [value];
  }
  content(text, raw = text) {
    if (this.#rawContentIndices.length == 0) { this.#rawContentIndices.push(this.#rawValues.length) }
    this.#rawValues.push(raw);
    this.#content.push(text);
  }
  subNode(tag, node) {
    this.#rawValues.push(node);
    let target = (this.#subNodes[tag] ||= []);
    target.push(node);
  }
  addRaw(str, isContentEnd = false) {
    if (isContentEnd) { this.#rawContentIndices.push(this.#rawValues.length); }
    this.#rawValues.push(str);
  }
  headerDone(raw) {
    this.#rawValues.push(raw);
    this.#rawContentIndices.push(this.#rawValues.length);
  }
  getContent() {
    switch (this.#contentMode) {
      case 'ARRAY': return this.#content.map(UniversalNode.realValue);
      case 'TEXT': return this.#content.join('').trim();
    }
    if (this.isMixed()) { return this.#rawValues.slice(...this.#rawContentIndices).join('').trim(); }
    else if (this.#content.length == 1) { return UniversalNode.realValue(this.#content[0]); }
    else { return this.#content.join('').trim(); }
  }

  valueOf() {
    let sanitizedObject = {};
    if (this.hasContent()) {
      if (!this.isMixed()) { return this.getContent(); }
      else { sanitizedObject['#text'] = this.getContent(); }
    }
    Object.keys(this.#subNodes).forEach(k => {
      if (this.#subNodes[k].length == 1) { sanitizedObject[k] = UniversalNode.realValue(this.#subNodes[k].at(0)); }
      else { sanitizedObject[k] = this.#subNodes[k].map(UniversalNode.realValue); }
    });
    let result = Object.assign({}, sanitizedObject);
    if (Object.keys(result).length > 0) { return result; }
    return null;
  }

}

/** @endsection */

/**
 * @section INTERNAL SUPPORT CLASSES AND FUNCTIONS
 * @description
 * Most of the classes and functions following this block are for parsing yml files
 * as this is the most complex format.
 */
function readTextPropertyFile(confFile, dataLinesCallback, contentProvider = CONTENT_PROVIDER.DIRECT, ...providerArgs) {
  if (Array.isArray(confFile)) { return dataLinesCallback(confFile); }

  log.debug('TRY READING %s', confFile);
  if (existsSync(confFile)) {
    CURRENT_FILE = confFile;
    log.info('READ %s', confFile);
  }
  else { log.info('Data to parse is string...') }
  let failure;
  try {
    let content = contentProvider(confFile, ...providerArgs);
    return dataLinesCallback(content);
  }
  catch (e) { log.error(e); failure = e; }
  finally { CURRENT_FILE = null }
  if (failure) { throw failure; }
}

/** To be used as a tag function for building regexes dynamically. Backslashes can be used as in literal RegExp with `/.../`, no escaping needed. */
function r(constants, ...paras) {
  let raw = constants.raw.map(e => e + (paras.shift() || '')).join('');
  if (raw.startsWith('/')) {
    let regTerm = raw.lastIndexOf('/');
    let flags = raw.substring(regTerm + 1);
    return new RegExp(raw.substring(1, regTerm), flags)
  }
  return new RegExp(raw);
}

/** YAML FILES */
const SINGLE_QUOTED = /'(.*)'/;
const DOUBLE_QUOTED = /"(.*)"/;
function unquote(value) {
  let quoted;
  if ((quoted = SINGLE_QUOTED.exec(value)) != null) { value = quoted[1].replaceAll("\\'", "'") }
  else if ((quoted = DOUBLE_QUOTED.exec(value)) != null) { value = quoted[1].replaceAll('\\"', '"') }
  return value;
}

/** Encapsulates transformation of a sequence of raw string tokens into one single string instance, folded, trimmed and chomped where needed. */
class YamlString {
  #token;
  #style = 'FLOW'; //FLOW, (Block-)FOLD or (Block-)LITERAL

  constructor(parentNode) { this.parent = parentNode; }
  isEmpty() { return this.#token.trim().length == 0; }

  get shiftLevel() {
    if (this.#style == "FLOW") { return '+' }
    if (typeof this.indentation == "number") { return this.parent[YAML.NESTING] + this.indentation; }
    let m = null;
    if ((m = /(?<=^|\n)[ \t]*(?=\S)/.exec(this.#token)) != null) { return m[0].length; }

    return Math.max(0, ...this.#token.split('\n').map(l => l.length));
  }

  add(text) {
    this.#token ||= '';
    this.#token += text;
  }

  /*
   * Applies the configuration / string styles to convert the raw array into a string, then returns the result.  
   *
   * To do so, the content must first be concatened, then modified in 4 locations, in the order below:
   * 1. Remove 'structural' whitespace: Remove a number of blanks matching the indentation from the start of every line (a.k.a. 'shifting').
   * 2. FOLD/FLOW: Remove trailing whitespaces per line and reduce subsequent empty lines by one.
   * 3. Trailing whitespace AFTER content. Trimmed or chomped.
   * 4. LITERAL: Keep shifted pre-content empty lines, trim in other styles.
  */
  get() {
    if (this.isEmpty()) { return; }

    let result = this.#token.replaceAll(/\r/g, '');

    // 1. structural whitespace. In FLOW style, this includes ALL whitespace at a line's start
    // for other styles, it depends on indentation and block header (handled by `get shiftLevel`)
    let shiftLevel = this.shiftLevel.replace(/(\d+)/, '{$1}');
    result = result.replace(r`/(?<=\n)[ \t]${shiftLevel}(?=\S)/g`, '');

    // 2. Folding - the most confusing part of string handling
    // It's different between FLOW and (Block-)FOLD, and FOLD also has a list of execeptions when
    // strings are NOT folded (which are very confusing and contradictory)
    if (this.#style == "FLOW") {
      // first remove whitespace before and after content in FLOW
      result = result
        .replace(/(?<=^|\n)[ \t]+(?=\S)/g, '')
        .replace(/(?<=\S)[ \t]+(?=\n|$)/g, '');
    }
    if (this.#style != "LITERAL") {
      // now to the actual folding: convert singular \n between none-spaces to blanks
      result = result.replace(/(?<=\S)\n(?=\S)/g, ' ');
      // block folding has a set of special exceptions which can't fully be covered by regex on \n alone
      // the hardest one is \n and empty lines following an indented line, which are not trimmed at all
      // this will be handled by ADDING an additional \n at the right place (which will be stripped)
      // by the next step again
      if (this.#style == "FOLD") {
        result = result.replace(/((^|\n)[ \t]+.*?\S[ \t]*\n)(\s*\n)(?![ \t])/g, '$1\n$3');
      }
      // remove \n which is followed by empty lines
      result = result.replace(/(?<=\S)\n(\n+)(?!\s)/g, '$1');
    }
    return unquote(result);
  }
}

/**
 * Helper class that can be used to transport meta-information about the origin of a property.
 * Most use cases shouldn't care because of the valueOf implementation and just behave as if it was a string.
 * One notable exception is default initialisation by existence checks.
 */
class PropValue {
  constructor(val, source = CURRENT_FILE) {
    this.value = val;
    this.source = source;
  }

  toJSON() { return this.value }
  valueOf() { return this.value }
  toString() { return String(this.value) }
}
function handleValue(value) {
  try { return new PropValue(JSON.parse(value)); } catch (e) { }
  return new PropValue(unquote(value));
}

/** @endsection */

let FORMATS = Object.keys(PARSE_FUNCTIONS);

module.exports = {
  parseDict, analyseProperty,
  SOURCE_FILE,
  // expose the parse functions themselves for places where the type is know, but not matching the extension
  confToDict, xmlToDict, xmlNew, yamlToDict, jsonToDict, esSettingsToDict, yamlToDict_IMPL,
  //for information/API puposes
  FORMATS,
  XML,
  PropValue
}
