// io/parsers.js
// SPDX-FileCopyrightText: 2025 Karsten Teichmann
//
// SPDX-License-Identifier: MIT

/**
 * @file
 * @description
 * This file contains a set of very simple parsers providing just enough of the specs/grammars 
 * to handle the different property file languages and styles used in batocera.linux:
 * - es_systems.yml
 * - es_features.yml
 * - configgen-*.yml
 * - *.conf files
 * - *.cfg files
 * - *.json
 * - *.xml
 * 
 * For XML and YAML formatted files, an external tool will be used (xq and yq) to transform to json first. 
 * This makes those types effectively twice as expensive to process.  
 * Thus, where speed matters, the user should not use yml. Moreover, frequently parsed files should be handled with 
 * dedicated in-process parsing functions.  
 * Example: `es_settings.cfg` is an xml file, but very simply structured.
 */
const log = require('logger').get()
const data = require('utils/data');

const path = require('node:path');
const { readFileSync, existsSync } = require('node:fs');
const { execSync } = require('node:child_process')

const SOURCE_FILE = Symbol.for('#SOURCE');
let CURRENT_FILE;

const PARSE_FUNCTIONS = Object.freeze({
  'yml': yamlToDict,
  'yaml': yamlToDict,
  'conf': confToDict,
  'json': jsonToDict,
  'cfg': esSettingsToDict,
  'xml': xmlNew,
  'amgp': xmlNew
});

const CONTENT_PROVIDER = Object.freeze({
  DIRECT: (filename, split = true) => {
    let result = existsSync(filename)
      ? readFileSync(filename, { encoding: 'utf8' })
      : filename;
    if (split) { result = result.split('\n') }
    return result;
  },
  YQ: (filename, split = true) => {
    return existsSync(filename)
      ? execSync(`yq . "${filename}"`, { encoding: 'utf8' })
      : execSync('yq', { encoding: 'utf8', input: filename });
  },
  XQ: (filename, split = true) => {
    return existsSync(filename)
      ? execSync(`xq . "${filename}"`, { encoding: 'utf8' })
      : execSync('xq', { encoding: 'utf8', input: filename });
  }
});

/**
 * This function is the main entrance to this file.  
 * It tries to auto-detect the parser function to use depending on the file type.  
 * The detection is rudimentary, it goes by extension. Concrete handler functions are looked up
 * in a constant object named `PARSE_FUNCTIONS`.
 * 
 * All handler functions follow the same contract:
 * - They will generally return nested object structures.
 * - Each actual property entry will we wrapped in an instance of `PropValue`.  
 * 
 * @exported
 * @see class PropValue
 */
function parseDict(confFile, typeOverride = null) {
  let fileType;
  let isPropertyString = false;
  if (existsSync(confFile)) {
    fileType = path.extname(confFile).substring(1);
  } else if (typeof confFile == "string") {
    fileType = 'conf';
    isPropertyString = true;
  }

  let effectiveType = typeOverride || fileType;
  let parser = PARSE_FUNCTIONS[effectiveType];

  if (isPropertyString) {
    //a string, but not a file path -> assume '.'-imploded property keys
    log.debug(`Parsing property string as [${effectiveType}] format`)
    return parser(confFile);
  }

  try {
    if (typeof parser == "undefined") { throw new Error('unsupported file type/extension: ' + confFile) }
    let resultDict = parser(confFile);
    resultDict[SOURCE_FILE] = confFile;
    return resultDict;
  } catch (e) {
    log.error("Could not parse file:", confFile, e);
    return {}
  }
}

/**
 * Handles `*.conf` files like `batocera.conf`.  
 * Capable of unpacking batocera's special syntaxes into regular tree structures:
 * - system["<game-name>"].something=...
 * - system.folder["<folder-path>"].something=...
 * 
 * @exported
 * @see analyseProperty
 */
function confToDict(confFile) {
  let assign = data.deepAssign;
  return readTextPropertyFile(confFile, lines => {
    let result = {};
    for (let line of lines) {
      let propLine = line.trim();
      if (propLine.startsWith('#') || !propLine.includes("=")) { continue; }
      let details = analyseProperty(propLine);
      assign(result, details.effectiveKey, details.value);
    }
    return result;
  })
}

const XML = {
  ENCODED_CHARS_REGEX: /&lt;|&gt;|&amp;|&apos;|&quot;/g,
  removeComments(lines) {
    if (Array.isArray(lines)) { lines = lines.join('\n') }

    while (lines.includes('<!--')) {
      let startIndex = lines.indexOf('<!--');
      let endIndex = lines.indexOf('-->', startIndex);
      lines = lines.substring(0, startIndex) + lines.substring(endIndex + 3);
    }
    return lines.split('\n');
  },

  condenseWhitespace(value) {
    return value.replace(/\s*$\s*/gm, ' ').trim();
  },

  decodeValue(value) {
    return value.replace(XML.ENCODED_CHARS_REGEX, match => {
      switch (match) {
        case '&lt;': return '<'
        case '&gt;': return '>'
        case '&amp;': return '&'
        case '&apos;': return "'"
        case '&quot;': return '"'
      }
    });
  },

  decodeAttribute(value) {
    return XML.decodeValue(XML.condenseWhitespace(value));
  },
}

const CFG_PROP_LINE = /<(\w+) name="(.*)" value="(.*)"\s*\/>/
const SETTINGS_TYPES = {
  string: String,
  float: Number,
  int: Number,
  bool: Boolean
}
function esSettingsToDict(cfgFile) {
  return readTextPropertyFile(cfgFile, (lines) => {
    lines = XML.removeComments(lines);
    let result = {};
    lines.map(_ => CFG_PROP_LINE.exec(_)).filter(_ => _ != null).forEach(line => {
      let key = XML.decodeValue(line[2]);
      let convertedProperty = `${key}=${XML.decodeValue(line[3])}`;
      let details = analyseProperty(convertedProperty);
      details.effectiveKey.set(result, SETTINGS_TYPES[line[1]](details.value.valueOf()));
    });
    return result;
  });
}

function jsonToDict(jsonFile, contentProvider = CONTENT_PROVIDER.DIRECT) {
  let isNativeJson = contentProvider == CONTENT_PROVIDER.DIRECT;

  function noComment(line) { return !/\s*\/\/.*/.test(line) }
  function propertyNodeCreator(key, value) {
    if (typeof value == "object" && value != null) { return value }
    else { return handleValue(String(value)) }
  }
  return readTextPropertyFile(jsonFile, (lines) => {
    let data = isNativeJson
      ? lines.filter(noComment).join('\n')
      : lines;
    try {
      return JSON.parse(data, propertyNodeCreator);
    } catch (e) {
      log.error(`Parsing [${jsonFile}] failed. Content was:`, data);
      log.error(e);
      return {};
    }
  }, contentProvider, isNativeJson);
}

function xmlToDict(xmlFile) {
  let result = jsonToDict(xmlFile, CONTENT_PROVIDER.XQ);
  return typeof result['NO-ROOT'] != "undefined"
    ? result['NO-ROOT']
    : result;
}

function xmlNew(xmlFile) {
  let result = readTextPropertyFile(xmlFile, sourceText => {

    let context = new ParseContext();
    context.addAttributeNode = function(m) {
      this.enterNode(`@${m[1]}`, m[1] + '=');
      this.addContent(handleValue(XML.decodeAttribute(m[2].slice(1, -1))), m[2])
      this.leaveNode();
    }.bind(context);

    let R = new RegexDocParser({
      // none-node whitespace
      BLK: [/\s+/ms, m => context.current.addRaw(m[0])],
      // start of node
      SON: [/<(?!\/)([\p{L}\._\d-]+?)([\s>]|(?=\/))/su, m => context.enterNode(m[1], m[0])],
      //attribute
      NATTR: [/(\w+)=("[^"]*"|'[^']*')/s, context.addAttributeNode],
      // end of node opening tag
      EOH: [/\s*\>/ms, m => context.nodeHeaderDone(m[0])],
      // node value/content
      VAL: [/(\S[^\<\>]*?)\s*(?=\<)/ms, m => context.addContent(handleValue(XML.decodeValue(m[1])), m[0])],
      // end of the node (closing tag)
      EON: [/<?\/[\p{L}\._\d-]*?>/msu, m => context.leaveNode(m[0])],
    }, 'XML');

    sourceText = sourceText
      .replace(/\<\?.*?\?\>/gs, '')
      .replace(/\<!--.*?--\>/gs, '');

    context.start(R);
    R.exec(sourceText);

    return context.current.valueOf();
  }, CONTENT_PROVIDER.DIRECT, false);
  return typeof result['NO-ROOT'] != "undefined"
    ? result['NO-ROOT']
    : result;
}

function yamlToDict(yamlFile) {
  return jsonToDict(yamlFile, CONTENT_PROVIDER.YQ);
}

function yamlToDict_IMPL(yamlFile) {
  return readTextPropertyFile(yamlFile, YAML.parse, CONTENT_PROVIDER.DIRECT, false);
}

const FOLDER_SPEC = /^\w+(\.folder\[(".*?")\/?\]).*=.*/;
const GAME_SPEC = /^\w+\[(".*?")\].*/;
function analyseProperty(propLine) {
  let fspath = [];
  if (fspath = FOLDER_SPEC.exec(propLine)) {
    propLine = propLine.replace(fspath[1], '');
    fspath = ['folder', unquote(fspath[2])];
  } else if (fspath = GAME_SPEC.exec(propLine)) {
    fspath = fspath[1];
    propLine = propLine.replace(`[${fspath}]`, '');
    fspath = ['game', unquote(fspath)];
  } else {
    fspath = [];
  }

  let equalPos = propLine.indexOf('=');
  let realKey = propLine.substring(0, equalPos).trim().split('.');
  let effKey = [...realKey]

  return {
    effectiveKey: new data.HierarchicKey(effKey.shift(), ...fspath, ...effKey),
    overridesKey: new data.HierarchicKey(...realKey),
    value: handleValue(propLine.substring(equalPos + 1).trim())
  }
}

/**
 * @section Generic parser infrastructure
 * @description
 * The following functions and classes provide basic plumbing and utility for all type-specific parser implementations.
 */

/**
 * A configurable generic parser based on regular expressions.  
 * The expressions will be set up to be sticky and share the same value for `lastIndex`.
 * Any successful match of one of the sub-expressions will advance the shared state,
 * but only when the match starts at the current `lastIndex`. It is not allowed to implicitely
 * skip and ignore parts of the input.
 * This allows to parse a long text without having to manipulate the input all the time 
 * which greatly improves speed. There is also no need to split line by line.  
 * 
 * Usage:
 * ```js
 * let ctx = new ParseContext();
 * let p = new RegexDocParser({
 *|  name : [ /pattern/flags, matchFunction(m){ ctx.doSomething(m) ... } ]
 * }, parserName)
 * 
 * p.exec(fullSource);
 * 
 * ```
 */
class RegexDocParser extends RegExp {
  #id;
  #log
  #state;
  #rules = [];
  #ruleIds = {};
  static #patchFlags(reg, flags) {
    flags ||= reg.flags;
    if (!flags.includes('y')) { flags += 'y'; }
    if (!flags.includes('d')) { flags += 'd'; }
    return flags;
  }
  static #escapeSpaces(str) { return str.replace(/\n/g, '\\n').replace(/\t/g, '\\t'); }
  static #execRule(reg, action, useAllowed, sourceText) {
    this.currentRule = reg;
    if (!useAllowed()) { return false; }
    let m = reg.exec(sourceText, false);
    if (m == null) { return false; }
    // skip empty matches
    if (m.indices[0][1] - m.indices[0][0] == 0) { return false; }

    let result = action
      ? action(m, reg, sourceText) || true
      : true;
    return result;
  }
  static #traverseSource(shared, sourceText) {
    let ruleIdx = 0;
    let lastRule = shared.#rules.length - 1;
    while (shared.lastIndex < sourceText.length) {
      let ruleResult = shared.#rules[ruleIdx](sourceText);
      shared.currentRule = null;
      if (ruleResult) { ruleIdx = shared.#ruleIds[ruleResult] || 0; }
      else { ruleIdx++; }

      if (ruleIdx > lastRule) { throw 'No valid expression for [' + this.snippet(sourceText) + ']'; }
    }
  }

  #initParser(conf, id) {
    this.#id = id || RegexDocParser.name;
    this.#log = require('logger').get(id);
    this.#log.excludeRules = [];
    Object.keys(conf).forEach(k => {
      if (this[k]) { throw `Duplicate pattern name: ${k} (or reserved word)`; }
      this[k] = new RegexDocParser(this, k, conf[k][0]);
      this.#ruleIds[k] = this.#rules.length;
      this.#rules.push(RegexDocParser.#execRule.bind(this, this[k], conf[k][1], conf[k][2] || (() => true)));
    });
    this.exec = function exec(str) { return RegexDocParser.#traverseSource(this, str); }.bind(this);
    this.lastIndex = 0;
    /* Not a real rule, it's placed like this to get access to the uniform log methods */
    this.stack = new RegexDocParser(this, '%STACK', /.*/);
  }

  #initSubRule(parent, id) {
    this.#id = id;
    this.#state = parent;
    this.#state.reset ||= this.reset.bind(this);
    this.#log = this.#state.#log;
  }

  constructor(state, id, a, b) {
    if (arguments.length <= 2) {
      super(/.*/);
      this.#initParser(state, id);
      return this;
    }

    super(a, RegexDocParser.#patchFlags(a, b));
    this.#initSubRule(state, id);
  }

  getId() { return this.#id; }
  reset() { this.#state.lastIndex = 0; }

  exec(sourceText) {
    let currentLast = this.#state.lastIndex || 0;
    this.lastIndex = currentLast;
    let match = super.exec(sourceText);

    if (match == null) { return null; }
    this.logRule(
      () => this.snippet(sourceText, currentLast)
        + '=> [' + RegexDocParser.#escapeSpaces((match || []).join('|')) + ']'
    )
    if (match.indices[0][0] != currentLast) {
      this.logDetails(() => "SKIP: did not start at lastIndex=" + currentLast, ", but at", this.lastIndex);
      return null;
    }

    this.#state.lastIndex = this.lastIndex;
    return match;
  }

  /** @returns {RegexDocParser} */
  get activeRule() {
    if (this.#state) { return this; }
    return this.currentRule || this;
  }
  #conditionalLog(level, suffix, ...data) {
    let id = this.#state ? this.#id : null;
    if (this.#log.excludeRules.includes(id || '%SELF')) { return; }
    let prefix = id ? `[${id.padStart(6)}] ` : '';
    if (suffix) { prefix += (suffix + ' '); }
    let handler = typeof data[0] == "function" ? data[0] : () => data.join(' ');
    this.#log.logWhen(level, () => prefix + handler());
  }

  /** 
   * An array of rule ids, and/or special tags '%SELF' and '%STACK'.  
   * Logs using `logRule` or `logDetails` from a rule will be supressed if a rule's id is contained in the exclude.  
   * Partial overriding is possible by using `log` instead, but this can in turn be disabled by excluding `%SELF`.  
   */
  setLogExcludes(f = []) { this.activeRule.#log.excludeRules = f; }
  /** Logs on INFO level and bound to `%SELF` topic instead of the rule id. */
  log(...data) { (this.#state || this).#conditionalLog('INFO', '', ...data); }
  /** Logs on DEBUG level */
  logRule(...data) { this.activeRule.#conditionalLog('INFO', '', ...data); }
  /** Logs on TRACE level  */
  logDetails(...data) { this.activeRule.#conditionalLog('INFO', '   -->', ...data); }

  /** Return a short part from the beginning of the given string, with breaks and tabs encoded. */
  snippet(sourceText, startIndex) {
    if (typeof startIndex != "number") { startIndex = this.lastIndex; }
    let snippetRaw = sourceText.substring(startIndex, startIndex + 30);
    return RegexDocParser.#escapeSpaces(snippetRaw).substring(0, 32).padEnd(35, '.');
  }
  toString() { return super.toString(); }
}

/**
 * Utilities to handle the complex parser state when parsing yaml.  
 * Methods of this object will be called with `this` set to the instance of the current `ParseContext`.
 */
class ParseContext {
  #STACK = []

  get root() { return this.#STACK[0]; }
  get current() { return this.#STACK.at(-1); }
  get previous() { return this.#STACK.at(-2); }

  /** 
   * `this` is normally lost when methods are used as callbacks.
   * This method overrides that by assigning a bound variant of the class method as instance property.
   * Passing the instance property as callback means passing a pre-bound function.
   */
  bindFunctions() {
    Object.values(Object.getOwnPropertyDescriptors(Object.getPrototypeOf(this))).forEach(v => {
      v = v.value;
      if (typeof v == "function") { this[v.name] = v.bind(this) }
    });
  }

  start(parser) {
    if (!parser) { throw 'parser argument required!'; }
    this.parser = parser;
    this.#STACK = [new UniversalNode('#root')];
  }

  enterNode(name, raw = name) {
    let node = new UniversalNode(name, raw);
    this.current.subNode(name, node);
    this.#STACK.push(node);
    this.parser.stack.logDetails('ENTER:', this.current.tagName);
  }

  addChildNode(tag) {
    let node = new UniversalNode(tag);
    this.current.content(node)
    this.#STACK.push(node);
    this.parser.stack.logDetails('PUSH CHILD:', tag)
  }

  nodeHeaderDone(raw) { this.current.headerDone(raw); }

  // leave node
  leaveNode(raw) {
    this.parser.stack.logDetails('LEAVE:', this.current.tagName)
    this.current.addRaw(raw, true);
    this.#STACK.pop();
  }

  addContent(text, raw) {
    this.parser.stack.logDetails(() => 'CONTENT: ' + JSON.stringify(text));
    this.current.content(text, raw);
  }
}

/**
 * To be used in conjunction with [class RegexDocParser](#class_regexdocparser).
 * 
 * This class represents one node of any type in the parsed tree. Its basic capabilities are inspired by the structure produced by `xmlToDict`.
 * Depending on how it is populated during parsing, its `valueOf()` method will return different objects:
 * Either simple scalar values, dictionaries or arrays.  
 * This implementation will also recurse into children and subNodes to ensure proper mapping of the entire graph.
 *
 * `UniversalNode` internally manages 3 containers where content can be added:
 * - `#subNodes = {}`: Populated by `attr(name, value)` and `subNode(tag, node)`.
 *   Used for key-value tree structures. Like in `xmlToDict`, the same key can be used multiple times, `#subNodes` maintains each key's containt as array.
 * - `#content = []`: For lists or otherwise mixed content which is not (or shall not) be added as a property of this instance.
 * - `#rawValues = []`: This is used in all content adding methods in addition to either `#subNodes` or `#content`
 *   to maintain an unmodified view of the literal input source. Required for some edge cases of mixed content (specific to XML).
 *
 * Result:  
 * Usually, only one of either `#subNodes` or `#content` will be populated and `UniversalNode` is able to decide on its own what to return.  
 * Refer to [valueOf()](#class_universalnode_valueof) for details.
 */
class UniversalNode {
  #rawValues = []
  #rawContentIndices = []
  #subNodes = {}
  #content = []
  #contentMode = 'DYNAMIC';

  static realValue(obj) { return (obj instanceof UniversalNode) ? obj.valueOf() : obj; }

  constructor(tag, raw) {
    this.tagName = tag;
    this.#rawValues.push(raw);
  }

  /** 
   * ContentMode allows to control how the children and subnodes of this will be added when `getContent()` is used.  
   * Must be one of DYNAMIC, ARRAY or TEXT.
   */
  setContentMode(cm) { this.#contentMode = cm; }
  getContentMode() { return this.#contentMode; }

  // some small getter utils
  hasContent() { return this.#content.length > 0; }
  getContentLength() { return this.#content.length; }
  isMixed() { return this.hasContent() && Object.keys(this.#subNodes).length > 0 }
  toString() { return this.#rawValues.join(''); }

  // methods to manipulate the content of the instance
  /** 
   * Add an 'attribute' to the node. Attributes are similar to subNodes, with 2 differences:
   * - The name given will automatically be prefixed with `@`.
   * - Attributes do not allow several items. Every assignment to the same attribute will overwrite the current value.
   */
  attr(name, value, raw) {
    this.#rawValues.push(raw);
    this.#subNodes[`@${name}`] = [value];
  }
  content(text, raw = text) {
    if (this.#rawContentIndices.length == 0) { this.#rawContentIndices.push(this.#rawValues.length) }
    this.#rawValues.push(raw);
    this.#content.push(text);
  }
  subNode(tag, node) {
    this.#rawValues.push(node);
    let target = (this.#subNodes[tag] ||= []);
    target.push(node);
  }
  addRaw(str, isContentEnd = false) {
    if (isContentEnd) { this.#rawContentIndices.push(this.#rawValues.length); }
    this.#rawValues.push(str);
  }
  headerDone(raw) {
    this.#rawValues.push(raw);
    this.#rawContentIndices.push(this.#rawValues.length);
  }
  getContent() {
    switch (this.#contentMode) {
      case 'ARRAY': return this.#content.map(UniversalNode.realValue);
      case 'TEXT': return this.#content.join('').trim();
    }
    if (this.isMixed()) { return this.#rawValues.slice(...this.#rawContentIndices).join('').trim(); }
    else if (this.#content.length == 1) { return UniversalNode.realValue(this.#content[0]); }
    else { return this.#content.join('').trim(); }
  }

  valueOf() {
    let sanitizedObject = {};
    if (this.hasContent()) {
      if (!this.isMixed()) { return this.getContent(); }
      else { sanitizedObject['#text'] = this.getContent(); }
    }
    Object.keys(this.#subNodes).forEach(k => {
      if (this.#subNodes[k].length == 1) { sanitizedObject[k] = UniversalNode.realValue(this.#subNodes[k].at(0)); }
      else { sanitizedObject[k] = this.#subNodes[k].map(UniversalNode.realValue); }
    });
    let result = Object.assign({}, sanitizedObject);
    if (Object.keys(result).length > 0) { return result; }
    return null;
  }

}

/** @endsection */

/**
 * @section Internal support classes and functions
 * @description The code in this section completely generic.  
 * It is used by the generic parser infrastructure, but also by other, simpler parser implementations.
 */

/**
 * Main entrance to all parser functions. It provides top-level handling of:
 * - Basic debug logging
 * - Handling of global state `CURRENT_FILE` for PropValue. Parsers depending on this are not 'thread-safe'.
 * - Errors: log exceptions and throw
 */
function readTextPropertyFile(confFile, dataLinesCallback, contentProvider = CONTENT_PROVIDER.DIRECT, ...providerArgs) {
  if (Array.isArray(confFile)) { return dataLinesCallback(confFile); }

  log.debug('TRY READING %s', confFile);
  if (existsSync(confFile)) {
    CURRENT_FILE = confFile;
    log.info('READ %s', confFile);
  }
  else { log.info('Data to parse is string...') }
  let failure;
  try {
    let content = contentProvider(confFile, ...providerArgs);
    return dataLinesCallback(content);
  }
  catch (e) { log.error(e); failure = e; }
  finally { CURRENT_FILE = null }
  if (failure) { throw failure; }
}

/**
 * To be used as a tag function for building regexes dynamically.  
 * Backslashes can be used as in literal RegExp with `/.../`, no escaping needed. 
 */
function r(constants, ...paras) {
  let raw = constants.raw.map(e => e + (paras.shift() || '')).join('');
  if (raw.startsWith('/')) {
    let regTerm = raw.lastIndexOf('/');
    let flags = raw.substring(regTerm + 1);
    return new RegExp(raw.substring(1, regTerm), flags)
  }
  return new RegExp(raw);
}

/** Used by `unquote` */
const QUOTE_MATCHER = /^\s*(["'])(.*)\1\s*$/s
/** 
 * Unquotes any **content** string enclosed in either `"` or `'` and returns the true inner string value.  
 * By default, the character used to as quotation is also unmasked (if it is by an uneven number of \).  
 * Quote unmasking can be disabled by passing false as second parameter.
 */
function unquote(value, unmask = true) {
  let matched = QUOTE_MATCHER.exec(value);
  if (matched == null) { return value; }
  if (unmask) { return matched[2].replace(r`(?<=^|[^\\](?:\\\\)*)\\${matched[1]}`, matched[1]); }
  else { return matched[2]; }
}

function handleValue(value) {
  try { return new PropValue(JSON.parse(value)); } catch (e) { }
  return new PropValue(unquote(value));
}

/**
 * Helper class that can be used to transport meta-information about the origin of a property.
 * Most use cases shouldn't care because of the valueOf implementation and just behave as if it was a string.
 * One notable exception is default initialisation by existence checks.
 */
class PropValue {
  constructor(val, source = CURRENT_FILE) {
    this.value = val;
    this.source = source;
  }

  toJSON() { return this.value }
  valueOf() { return this.value }
  toString() { return String(this.value) }
}

/** @endsection */

/**
 *  @section YAML-specific code
 */
class YAML extends ParseContext {
  static BLOCK_H_STYLES = { '|': "LITERAL", '>': "FOLD" }
  static CHOMP_MODES = { '+': 'KEEP', '-': 'STRIP', '': 'CLIP' }

  static CNEST = Symbol.for('CNEST');
  static NESTING = Symbol.for('NESTING');
  /** BLOCK or FLOW */
  static MODE = Symbol.for('MODE');
  /** ENTRY, TOKEN, TOKEN:ML */
  static PHASE = Symbol.for('PHASE');
  static TOKEN = Symbol.for('TOKEN');

  static parse(sourceText) {
    let yaml = new YAML();
    yaml.bindFunctions();

    function isMode(m, isIncluded = true) { return () => m.includes(yaml.current[YAML.MODE]) == isIncluded; }
    function isPhase(p, isIncluded = true) { return () => p.includes(yaml.current[YAML.PHASE]) == isIncluded; }

    let R = new RegexDocParser({
      // NOOP: Comments - (at least) one whitespace followed by whatever up to the next \n.
      // only allowed while not in the middle of a multiline token, treat it as regular content otherwise
      LC: [/(?:^|\n[ \t]*|[ \t]*)[#%].*?(?=$|\n)/, false, isPhase(["TOKEN:ML"], false)],
      // Empty lines
      EL: [/(\n[ \t]*)(?=\n|$)/, yaml.handleToken],
      'BL:ARR': [/(-)(?=\s)/, yaml.markArray, isMode('BLOCK')],
      // From " on anything up until the first unescaped " (unescaped: no uneven number of slashes before it)
      // can run over multiple lines.
      DQ: [/("(?:\\\\|\\"|\\[^"]|[^\\"])*")/m, yaml.handleToken],
      // single quoted tokens
      SQ: [/('(?:''|[^'])*')/m, yaml.handleToken],
      // block scalar indicators
      'BL:SLH': [/([|>](?:[+-]\d?|\d[+-]?)|[|>])(?=[ \t\n])/, yaml.handleOperator, isMode('BLOCK')],
      // indicators which are not allowed in unquoted tokens
      OP: [/([:?](?=[ \t\n]))/, yaml.handleOperator],
      // indicators/operators belonging to flow mode: [], {}, ','
      'FL:OP': [/\{|\[/, yaml.enterFlow],
      'FL:X': [/\s*(?:,[ \t]*\}|,[ \t]*\]|[,}\]])/, yaml.closeFlow, isMode('FLOW')],
      // unquoted in FLOW mode has additional terminators `,`, `}` and `]`
      'NQ:FL': [/(\S.*?)(?=[:?-]\s|[\]},]|\s#|\n)/, yaml.handleToken, isMode('FLOW')],
      // any unquoted one-liner, up to the first indicator or linebreak
      // note: might have issues at end of file - requires an additional space or \n
      NQ: [/(\S.*?)(?=[:?-]\s|\s#|\n|$)/, yaml.handleToken],
      // the whitespace at the beginning of a line which controls when to leave a nested level
      DPT: [/(?:^|\n)([ \t]*)(-\s)?(?=[^\-\n]|\S)/, yaml.closeNodes, isMode('BLOCK')],
      // In FLOW, only take the leading whitespaces as parts of the token, they are not used for closing.
      'DPT:FL': [/(\n\s*)/, yaml.handleToken, isMode('FLOW')],
      // NOOP: trailing whitespace from payload up to the next \n
      TRAIL: [/([ \t]+)(?=\S|\n|$)/m, yaml.handleToken]
    }, 'YAML');

    R.setLogExcludes(['TRAIL']);
    yaml.start(R);
    yaml.current[YAML.NESTING] = -1;
    yaml.current[YAML.MODE] = 'BLOCK';
    yaml.current[YAML.PHASE] = 'ENTRY';

    R.exec(sourceText);
    yaml.closeNodes(["", ""]);
    yaml.cleanNode();

    return yaml.root.valueOf();
  }

  cleanNode() { Object.getOwnPropertySymbols(this.current).forEach(k => delete this.current[k]); }
  pushNextNode(tagName, isSubNode) {
    if (typeof isSubNode == "undefined") { isSubNode = tagName ? String(tagName).length > 0 : false; }
    let factory = isSubNode ? this.enterNode : this.addChildNode;
    factory.call(this, tagName);
    this.inheritStateProperties();
    this.current[YAML.PHASE] = "ENTRY";
    delete this.previous[YAML.PHASE];
  }
  inheritStateProperties() {
    this.current[YAML.MODE] = this.previous[YAML.MODE] || 'BLOCK';
    this.current[YAML.NESTING] = this.previous[YAML.CNEST] || 0;
    this.parser.logDetails(
      `INHERIT ${this.current.tagName}/${this.current[YAML.NESTING]} <-`,
      `${this.current[YAML.MODE]}:${this.previous.tagName}/${this.previous[YAML.CNEST]}`
    );
  }

  hasToken() { return typeof this.state(YAML.TOKEN) != "undefined"; }
  handleToken(match, rule = null) {
    // when coming from parser directly, the first group will be token content
    // otherwise it could be called from another method with a pre-processed string
    match = Array.isArray(match) ? match[1] : match;
    let token = (this.current[YAML.TOKEN] ||= new YamlToken(this));
    if (token.isEmpty() && rule != null && ['DQ', 'SQ'].includes(rule.getId())) { token.quoteMode = rule.getId(); }
    token.add(match);
    switch (this.state(YAML.PHASE)) {
      case "ENTRY":
        if (!token.isEmpty()) { this.state(YAML.PHASE, "TOKEN"); }
        break;
      case "TOKEN":
        this.state(YAML.PHASE, "TOKEN:ML");
        break;
    }
  }
  consumeToken() {
    if (!this.hasToken()) { return; }

    let result = this.state(YAML.TOKEN).get();
    delete this.current[YAML.TOKEN];
    delete this.current[YAML.PHASE];
    return result;
  }

  tokenToContent() {
    let token = this.consumeToken();
    if (typeof token == "undefined") { return }
    this.addContent(new PropValue(token));
  }

  enterFlow(flowNodeMarker) {
    this.pushNextNode('#FLOW:ROOT', false);
    this.current[YAML.MODE] = 'FLOW';
    if (flowNodeMarker[0].trim() == "[") { this.current.setContentMode("ARRAY"); }
    if (this.current.getContentMode() == "ARRAY") {
      this.pushNextNode('#ARR-ENTRY', false);
    }
  }

  closeFlow(match) {
    let term = match[0].replace(/\s+/gm, '');
    if (term == ',') {
      this.current.getContentMode() == "ARRAY" ? this.tokenToContent() : this.finishCurrentNode();
      return;
    }

    this.finishCurrentNode();
    if (!this.previous) { return; }
    if (this.current.tagName == "#FLOW:ROOT") { this.finishCurrentNode(); }
  }

  handleOperator(op) {
    op = op[0];
    switch (op.charAt(0)) {
      case ':':
        this.pushNextNode(this.consumeToken());
        break;
      case '|':
      case '>':
        let tokenStyle = YAML.BLOCK_H_STYLES[op.charAt(0)];
        let token = new YamlToken(this, tokenStyle);
        token.parser = this.parser;
        this.state(YAML.TOKEN, token);
        token.chomping = YAML.CHOMP_MODES[op.replace(/[^+-]/g, '')];
        token.indentation = op.replace(/[^\d]/g, '');
    }
  }

  markArray(depth, addChild = true) {
    depth = Number(depth) || 2;
    if (this.current.getContentMode() != "ARRAY") {
      this.pushNextNode(this.current.tagName + '[]', false);
      this.state(YAML.CNEST, this.previous[YAML.CNEST]);
      this.current.setContentMode("ARRAY");
    }

    if (addChild) {
      this.pushNextNode(this.current.tagName + '#' + this.current.getContentLength(), false);
      this.state(YAML.CNEST, this.state(YAML.NESTING) + depth);
    }
  }

  finishCurrentNode() {
    this.parser.logDetails('FINISH:', this.current.tagName);
    this.tokenToContent();
    this.cleanNode();
    this.leaveNode();
  }

  state(state, value, node) {
    node ||= this.current;
    if (value == undefined) { return node[state]; }
    node[state] = value;
    this.parser.logDetails(`${Symbol.keyFor(state)} ${value} >>> ${node.tagName}`);
  }

  /** Only set value if state is not set already */
  stateOnce(state, value, node) {
    node ||= this.current;
    if (typeof value == "undefined") { throw "stateOnce() must receive a value argument" }
    if (typeof node[state] == "undefined") { this.state(state, value, node); }
  }

  closeNodes(depthMatch, rule) {
    let regD = depthMatch[1].length;
    let arrD = regD;
    if (depthMatch[0].trim().endsWith('-')) {
      arrD += 2;
      this.parser.lastIndex -= 2;
    }
    while (true) {
      if (arrD > regD && arrD > (this.state(YAML.NESTING) || 0) && "ENTRY" == this.state(YAML.PHASE)) {
        this.state(YAML.CNEST, regD);
        this.markArray(arrD - regD, false);
      }
      let c = this.current;
      let diff = this.state(YAML.NESTING) - (c.getContentMode() == "ARRAY" ? arrD : regD);
      this.parser.logDetails(
        `RCL:${c.tagName} N:${c[YAML.NESTING]} D:${diff} (-${c.getContentMode() == "ARRAY" ? arrD : regD})`
      );

      if (diff >= 0) { this.finishCurrentNode(depthMatch); }
      else { break; }
    }
    this.stateOnce(YAML.CNEST, arrD);
    //save whitespace for unquoted multiline Strings
    this.handleToken(depthMatch[0].substring(0, regD + 1), rule);
  }
}

/** 
 * Encapsulates transformation of raw (multiline) tokens into the corresponding final output value.
 * - It handles all folding, trimming and chomping for styles of strings.
 * - One-word unquoted tokens will be evaluated as JSON primitives with string as fallback on failure.
 *
 * **API:**  
 * 1. token style is set in constructor, must be one of (`FLOW`, (Block-)`FOLD` or (Block-)`LITERAL`)
 * 2. `YamlToken.quoteMode` is for flow and must be one of (`DQ`, `SQ` or `NQ`).
 * 3. (optional) `YamlToken.indentation` and `YamlToken.chomping` (`STRIP`, `CLIP` or `KEEP`) are for values coming from block headers.
 * 4. Incrementally add raw input with `YamlToken.add(str)`. For flow tokens, the quote characters **must** be part of the raw input.
 * 5. When done, retrieve the value with `YamlToken.get()`.
 */
class YamlToken {
  static JSON_MASKING = { '\n': '\\n', '\t': '\\t' }

  #style = 'FLOW'; //FLOW, (Block-)FOLD or (Block-)LITERAL
  #quoting = 'NQ';
  #chompMode = 'CLIP';
  #indent;

  #token = '';
  #trailing;

  constructor(context, style) {
    this.parser = context.parser;
    this.parent = context.current;
    this.#style = style || this.#style;
  }
  isEmpty() { return (typeof this.#token == "undefined") || this.#token.trim().length == 0; }

  /** @arg {"CLIP"|"STRIP"|"KEEP"} style */
  set chomping(style) { this.#chompMode = style; }
  /** @arg {"DQ"|"SQ"|"NQ"} mode sets quoting for "FLOW" style tokens */
  set quoteMode(mode = "NQ") { this.#quoting = mode; }
  /** @arg {string|number} raw */
  set indentation(raw) {
    let n = parseInt(raw);
    if (Number.isNaN(n) || n < 1) { this.#indent = undefined; }
    else this.#indent = n;
  }

  get shiftLevel() {
    if (this.#style == "FLOW") { return '+' }
    if (typeof this.#indent == "number") { return this.parent[YAML.NESTING] + this.#indent; }

    let m = null;
    // YAML 1.2.2 - 8.1.1.1 Block Indentation Indicator
    // 1. whitespace prefix of first none-empty line
    if ((m = /(?<=^|\n)[ \t]*(?=\S)/.exec(this.#token)) != null) { return m[0].length; }
    // 2. Longest empty line
    return Math.max(0, ...this.#token.split('\n').map(l => l.length));
  }

  add(text) {
    this.parser.logDetails(() => { return 'TOKEN-ADD: ' + JSON.stringify(text); });
    this.#token += text;
  }

  /**
   * Applies the configuration / string styles to convert the raw token into a yaml value, then returns the result.  
   *
   * To do so, the content must first be concatenated, then modified in 5 locations/steps, in the order below:
   * 1. Remove 'structural' whitespace: Remove a number of blanks from the start of every line (='shift').
   * 2. FOLD/FLOW: Remove trailing whitespaces per line and reduce subsequent empty lines by one (='squash').
   * 3. LITERAL: Keep shifted pre-content empty lines, trim in other styles.
   * 4. Trailing whitespace AFTER content: Trim or chomp.
   * 5. After sanitization, the content must be processed:
   *   a) When DQ -> JSON.parse to handle escapes
   *   b) SQ: remove outer '
   *   c) unquoted, one-word values: try JSON.parse for numbers/boolean, return as string on error.
   *   d) assume string for everything else
  */
  get() {
    this.parser.logDetails('TOKEN-GET:', this.#style, this.#quoting, this.shiftLevel, this.#chompMode, 'RAW:', JSON.stringify(this.#token));
    if (this.isEmpty()) { return; }

    // 1. Basic normalization and structural cleaning
    let result = this.#sanitizedToken();

    // 2. Folding - the most confusing part of string handling
    // It's different between FLOW and (Block-)FOLD, and FOLD also has a list of execeptions when
    // strings are NOT folded (which are very confusing and contradictory)
    if (this.#style == "FLOW") {
      result = this.#prepareFlow(result);
    } else if (result.charAt(0) == '\n') {
      // BLOCK headers must consume the very first `\n`, it is NOT part of the string.
      // This line assumes that the parser does not consume linebreaks when evaluating the header,
      // but rather includes them into whitespace when building the token (by pattern `DPT`).
      result = result.substring(1);
    }
    // 3. Squashing of empty lines
    if (this.#style != "LITERAL") { result = this.#handleFolding(result); }
    // 4. Chomp
    switch (this.#chompMode) {
      case 'CLIP': result += this.#trailing.length > 0 ? '\n' : ''; break;
      case 'KEEP': result += this.#trailing; break;
    }
    return this.#postProcess(result);
  }

  /**
   * Peforms 3 actions on `#token` and returns the result:
   * 1. Basic normalizations of encoding (`NFC`) and linebreaks (= drop `\r`)
   * 2. Drop a number of (structural) whitespace characters at the beginning of each line.  
   *    For flow tokens, all none-content spaces are dropped.  
   *    For blocks, it depends on the indentation (and header).  
   *    This step is elementary for the following logic, because it converts the yaml definition of an 'empty line'
   *    into a plain, solitary `\n`. Any space or tab left after this is **content**.  
   *    Subsequent logic won't have to deal with 'empty lines', so folding is mostly identical for FLOW and FOLD.
   * 3. **Empty** trailing newlines are removed and stashed away into `#trailing`
   */
  #sanitizedToken() {
    let result = this.#token.normalize('NFC').replaceAll(/\r/g, '');

    // Structural whitespace
    let shiftLevel = String(this.shiftLevel).replace(/(\d+)/, '{0,$1}');
    let shiftReg = r`/(?<=^|\n)[ \t]${shiftLevel}/g`;
    result = result.replaceAll(shiftReg, '');

    //stash and drop trailing lines from the last content char (last character which is NOT `\n`) to end of string
    this.#trailing = /(?<!\n)(\n*)?$/.exec(result)[1] || '';
    if (this.#trailing.length > 0) { result = result.slice(0, -this.#trailing.length); }

    return result;
  }

  /**
   * Do some preparation specific to `FLOW` tokens.  
   * Depending on quoting style:
   * - DQ: Strip escaped newlines
   * - SQ: Unescaped `''` by de-duplicating it.
   * - NQ: completely drop any whitespace (including newlines) **before** the content
   *
   * Always: Strip all leading and trailing whitespace (spaces and tabs) on each line.
   */
  #prepareFlow(result) {
    this.#chompMode = "STRIP";
    switch (this.#quoting) {
      // Drop 'escaped' `\n` - these are yaml helpers to break the source without adding LFs to content.
      // Other escape sequences will be dealt with by `JSON.parse(...)`,
      // which is why the " surrounding the string must not be removed.
      case 'DQ': result = result.replace(/\\\n/g, ''); break;
      // Un-escape singular content apostrophs.
      case 'SQ': result = result.replace(/''/g, "'"); break;
      // `trimStart()` is only necessary when not quoted, whitespace outside of quotes will be dropped later.
      // When the string doesn't start with a quote char, it's an unquoted token containing " or ' as content.
      case 'NQ': default: result = result.trimStart();
    }
    // YAML 1.2.2 - 6.5. Line Folding - Flow Folding:
    // remove whitespace before and after line content
    return result
      .replace(/(?<=^|\n)[ \t]+(?=\S)/g, '')
      .replace(/(?<=\S)[ \t]+(?=\n|$)/g, '');
  }

  /**
   * 1. Convert singular newlines to spaces
   * 2. Squash subsequent newlines (and 'empty' lines) where applicable (some exception for block in `FOLD` style)
   */
  #handleFolding(result) {
    // YAML 1.2.2 - (Block Scalar) 8.1.3. Folded Style
    // Deal with the none-folded exceptions by ADDING an additional `\n`, which will be stripped by the next step again.
    // This is easier to detect than excluding the exceptions from the generic line folding used below.
    if (this.#style == "FOLD") {
      //A line starting with at least one content space, plus (optionally) more empty lines, followed by none-space
      //-> deeper nested line + linebreaks + 'unnested' content
      result = result.replace(/(?<=^|\n)([ \t].*?\n)(\n*)(?=\S)/g, '$1\n$2');
    }
    // YAML 1.2.2 - 6.5. Line Folding
    return result
      // 1. Convert singular `\n` between content-characters to blanks (implicitely excludes the FOLD-exception `\n[space]`).
      // any character except newline can be content to the left of \n, none-space is required afterwards
      .replace(/(?<=[^\n])\n(?=\S)/g, ' ')
      // 2. Remove one `\n` which is followed by (at least one) empty line(s)
      .replace(/(?<!^)\n(\n+)(?![ \t])/g, '$1');
  }

  #jsonMask(string, subs = YamlToken.JSON_MASKING) {
    let exp = r`/(${Object.keys(subs).join('|')})/g`
    return string.replace(exp, (f) => subs[f])
  }

  /**
   * `FLOW` tokens are interpreted differently, based on their quoting style. This method takes care of that.
   * - `DQ` supports backslash escapes
   * - `SQ` are literal strings, but the outer ' must be stripped, as these are **not** content.
   * - `NQ` might be JSON primitives (booleans, numbers, null).
   */
  #postProcess(result) {
    // blocks of any kind are strings, there is no clever evaluation of content or handling of quotes
    if (this.#style != 'FLOW') { return result; }
    result = result.trim();
    switch (this.#quoting) {
      // If it has any whitespace, it's a string. If not, fallthrough as it might be another JSON primitive.
      case 'NQ': if (/\s/.test(result)) { return result; }
      // Double-quoted is parsed as JSON for backslash sequences, fallback to just unquoting on failure
      case 'DQ': try { return JSON.parse(this.#jsonMask(result)); } catch { /* ignore */ };
      // Single-quoted doesn't support escapes other than '' which is handled in `#prepareFlow()`, so just unquote
      case 'SQ': result = unquote(result, result.startsWith('"'));
    }
    return result;
  }
}

/** @endsection */

let FORMATS = Object.keys(PARSE_FUNCTIONS);

module.exports = {
  parseDict, analyseProperty,
  SOURCE_FILE,
  // expose the parse functions themselves for places where the type is know, but not matching the extension
  confToDict, xmlToDict, xmlNew, yamlToDict, jsonToDict, esSettingsToDict, yamlToDict_IMPL,
  //for information/API puposes
  FORMATS,
  XML,
  PropValue
}
